File: 94.binary-tree-inorder-traversal.go
/*
 * @lc app=leetcode id=94 lang=golang
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Example 1:

// Input: root = [1,null,2,3]

// Output: [1,3,2]

// Example 2:

// Input: root = [1,2,3,4,5,null,8,null,null,6,7,9]

//   1
// 2 3
//4 5 null 8
//null null 6 7 9
// Output: [4,2,6,5,7,1,3,9,8]

func inorderTraversal(root *TreeNode) []int {
	var result []int
	var traverse func(node *TreeNode)

	//Recursive function for traversal
	traverse = func(node *TreeNode) {
		//檢查該node有沒有值,沒有值就跳出,繼續做其他的traverse
		//如果是看Left,沒有值則append節點去result
		if node == nil {
			return
		}
		// Traverse the left subtree
		traverse(node.Left)
		// Visit the root node
		result = append(result, node.Val)
		// Traverse the right subtree
		traverse(node.Right)
	}

	//執行tree的遍歷
	traverse(root)
	return result
}

// @lc code=end

File: 94.binary-tree-inorder-traversal.py
#
# @lc app=leetcode id=94 lang=python
#
# [94] Binary Tree Inorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        # Recursive Approach
        # 中序遍歷: 左子樹 -> 根節點 -> 右子樹
        # Inorder Traversal: Left Subtree -> Root Node -> Right Subtree

        # 建立一個空的list來存放節點值
        # Create a list to store the node values
        result =[]

        def inorder(node):
            # 如果root為空,返回
            # If root is empty, return
            if not node:
                return
            # 先遍歷這個點的左子樹
            # Traverse the left subtree of this node first
            inorder(node.left)
            # 處理完左子樹後,先將節點值加入到result中
            # After traversing the left subtree, add the node value to result
            result.append(node.val)
            # 繼續遍歷右子樹
            # Traverse the right subtree
            inorder(node.right)
        # 從根部開始執行
        # Start from the root
        inorder(root)

        return result
        
# @lc code=end

